src/ir/auth_logic/universe_relation_insertion.cc

```diff
@@ -0,0 +1,357 @@
+      TypeToUniverseDeclaration(datalog::ArgumentType(
+          datalog::ArgumentType::Kind::kNumber, "Number")),
+      TypeToUniverseDeclaration(datalog::ArgumentType(
+          datalog::ArgumentType::Kind::kPrincipal, "Principal")),
+  };
+  ret.merge(builtin_universes);
+  return ret;
+}
+
+//----------------------------------------------------------------------------
+// Grounding Condition Transformation
+//----------------------------------------------------------------------------
+// This class transforms SaysAssertions so that if
+// there are any potentially ungrounded variables on the LHS of an assertion,
+// conditions are added to the RHS that check that this variable
+// is in a universe. As a way of safely overapproximating the ungrounded
+// variables, this just assumes any *non-constant* arguments in the RHS
+// are ungrounded, and adds conditions for them to the LHS.
+// It implements just one public method,
+// `AddUniverseConditions(SaysAssertion says_assertion)`
+// that returns a transformed version of the says assertion.
+// Internally, it extends a visitor for a part of this translation
+class UniverseConditionTransformer
+    : public AuthLogicAstTraversingVisitor<UniverseConditionTransformer,
+                                           std::vector<BaseFact>> {
+ public:
+  UniverseConditionTransformer(const DeclarationEnvironment& decl_env)
+      : decl_env_(decl_env) {}
+
+  // The `AddUniverseConditions` methods, including this one,
+  // are not implemented as actual visits because these methods
+  // are implementing a transformation, and the returned values
+  // need to be placed in a part of the syntactic context that
+  // is specific to each syntactic element.
+  SaysAssertion AddUniverseConditions(const SaysAssertion& says_assertion) {
+    std::vector<Assertion> assertion_list;
+    for (const Assertion& assertion : says_assertion.assertions()) {
+      assertion_list.push_back(AddUniverseConditions(assertion));
+    }
+    return SaysAssertion(says_assertion.principal(), std::move(assertion_list));
+  }
+
+ private:
+  Assertion AddUniverseConditions(const Assertion& assertion) {
+    return std::visit(
+        [this](const auto& variant) { return AddUniverseConditions(variant); },
+        assertion.GetValue());
+  }
+
+  Assertion AddUniverseConditions(const Fact& fact) {
+    std::vector<BaseFact> conditions = GetUniverseConditions(fact.base_fact());
+    if (conditions.size() > 0) {
+      return Assertion(ConditionalAssertion(fact, {conditions}));
+    } else {
+      return Assertion(fact);
+    }
+  }
+
+  Assertion AddUniverseConditions(const ConditionalAssertion& cond_assertion) {
+    std::vector<BaseFact> conditions =
+        GetUniverseConditions(cond_assertion.lhs().base_fact());
+    absl::c_copy(cond_assertion.rhs(), std::back_inserter(conditions));
+    return Assertion(ConditionalAssertion(cond_assertion.lhs(), conditions));
```

Comment:
Might want to break `cond_assertion.lhs()` into a temporary, just to make things a bit more concise.